pkg/controller/cache.go

```diff
@@ -0,0 +1,135 @@
+package controller
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+)
+
+// TopologyInfo holds the data for NodeLabels and TopologyKeys
+type TopologyInfo struct {
+	NodeLabels     map[string]string
+	TopologyKeys   []string
+	RequisiteTerms []topologyTerm
+}
+
+// TopologyProvider is an interface that defines the behavior for looking up
+// a TopologyInfo object by its pvc UID.
+type TopologyProvider interface {
+	GetByPvcUID(pvcUID types.UID) (*TopologyInfo, error)
+	// The entry is deleted when provision succeeds or returns a final error.
+	Delete(pvcUID types.UID) error
+
+	// Update methods now perform an "upsert" and don't return errors.
+	UpdateNodeLabels(pvcUID types.UID, newLabels map[string]string)
+	UpdateTopologyKeys(pvcUID types.UID, newKeys []string)
+	UpdateRequisiteTerms(pvcUID types.UID, requisiteTerms []topologyTerm)
+}
+
+// InMemoryStore is a concrete implementation of TopologyProvider.
+// It uses an in-memory map for quick lookups.
+type InMemoryStore struct {
+	// The map key is the object's name.
+	data map[types.UID]*TopologyInfo
+	// Adding a mutex for thread-safe access
+	mutex sync.RWMutex
+}
+
+// NewInMemoryStore creates and initializes a new store.
+func NewInMemoryStore() *InMemoryStore {
+	return &InMemoryStore{
+		data: make(map[types.UID]*TopologyInfo),
+	}
+}
+
+// Delete implements the TopologyProvider interface.
+// It uses the built-in delete() function to remove the item from the map.
+func (s *InMemoryStore) Delete(pvcUID types.UID) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	// First, check if the key exists to provide a helpful error.
+	_, found := s.data[pvcUID]
+	if !found {
+		klog.Warningf("cannot delete: object with pvcUID '%s' not found", pvcUID)
```

Comment:
do we need this? If something returns an error early, it's possible that nothing gets added to the cache

Reply:
Do we need signal at all if there is something wrong with the cache? Like somehow the cache is not written properly so it is missing from deletion.

Reply:
Removed